plugins/woocommerce/includes/rest-api/Controllers/Version2/class-wc-rest-coupons-v2-controller.php

```diff
@@ -304,6 +313,13 @@ protected function prepare_object_for_database( $request, $creating = false ) {
 						break;
 					default:
 						if ( is_callable( array( $coupon, "set_{$key}" ) ) ) {
+							// Special validation for amount when discount_type is percent.
+							if ( 'amount' === $key && 'percent' === ( $request['discount_type'] ?? 'fixed_cart' ) ) {
+								$amount = wc_format_decimal( $value );
+								if ( $amount > 100 ) {
+									return new WP_Error( 'woocommerce_rest_invalid_coupon_amount', __( 'Percentage discount cannot exceed 100%.', 'woocommerce' ), array( 'status' => 400 ) );
+								}
+							}
```

Comment:
This is where the main change was made (besides additional tests). The rest of the changes are from autoformatting fixes.